presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -373,6 +373,13 @@ QueryInfo getQueryInfo(Optional<StageInfo> rootStage)
 
         boolean completeInfo = getAllStages(rootStage).stream().allMatch(StageInfo::isFinalStageInfo);
         boolean isScheduled = isScheduled(rootStage);
+        List<TaskId> failedTasks = getAllStages(rootStage).stream()
```

Comment:
I will double check the frequency of `getBasicQueryInfo` get called, and whether this could be expensive (since it has to iterate all the tasks)

Reply:
`failedTasks` is only included in `QueryInfo` so I assume you mean `QueryInfo`. There are several scenarios that we need to iterate all the tasks:

- `ExplainAnalyzeOperator#getOutput`. Only once
- In query state change listener, see `Query#create`. This will be done once after query state changed to a "done" state.
- In `Query#getNextResult`, this will be called frequently
- In `QueryResource#getQueryInfo`, this could be called frequently from client side
- In `QuerySystemTable#cursor`, I honestly don't know how frequently this would be called.

I think we should only iterate all the tasks when the `QueryInfo` is finalized, because it is mainly used for `QueryCompletedEvent` and not used while query is running.